data_structures/arrays/product_sum.py

```diff
@@ -0,0 +1,71 @@
+"""
+Calculate Product Sum from an Array.
+reference: https://dev.to/sfrasica/algorithms-product-sum-from-an-array-dc6
+
+For doctests run following command:
+python -m doctest -v product_sum.py
+
+We need to create a function that calculates the product sum of a "special" array.
+This array can contain integers or nested arrays. The product sum is obtained by
+adding all the elements together and multiplying by their respective depth.
+
+For example, in the array [x, y], the product sum is (x + y). In the array [x, [y, z]],
+the product sum is x + 2 * (y + z). In the array [x, [y, [z]]],
+the product sum is x + 2 * (y + 3z).
+
+Example Input:
+[5, 2, [-7, 1], 3, [6, [-13, 8], 4]]
+Output: 12
+
+"""
+
+
+def product_sum(arr: list[int | list], depth: int) -> int:
```

Comment:
This algo has doc tests
`  Examples:
        >>> product_sum_array([1, 2, 3])
        6
        >>> product_sum_array([1, [2, 3]])
        11
        >>> product_sum_array([1, [2, [3, 4]]])
        47`

Reply:
This function has no doctests.  If we copy just this function and put it into another project, it will have zero test coverage.  That is not acceptable.

Reply:
Hi, I have added the doctests for product_sum function, can you please review